indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java

```diff
@@ -269,11 +302,90 @@ public void setConf(Configuration config)
     }
   }
 
-  public static class IndexGeneratorReducer extends Reducer<BytesWritable, Writable, BytesWritable, Text>
+  public static class IndexGeneratorCombiner extends Reducer<BytesWritable, BytesWritable, BytesWritable, BytesWritable>
+  {
+    protected HadoopDruidIndexerConfig config;
+    private AggregatorFactory[] combiningAggs;
+
+    @Override
+    protected void setup(Context context)
+        throws IOException, InterruptedException
+    {
+      config = HadoopDruidIndexerConfig.fromConfiguration(context.getConfiguration());
+
+      List<AggregatorFactory> combiningAggsList = Lists.newArrayList();
+      for (AggregatorFactory factory : config.getSchema().getDataSchema().getAggregators()) {
+        combiningAggsList.add(factory.getCombiningFactory());
+      }
+      combiningAggs = combiningAggsList.toArray(new AggregatorFactory[0]);
+    }
+
+    @Override
+    protected void reduce(
+        final BytesWritable key, Iterable<BytesWritable> values, final Context context
+    ) throws IOException, InterruptedException
+    {
+
+      Iterator<BytesWritable> iter = values.iterator();
+      BytesWritable first = iter.next();
+
+      if(iter.hasNext()) {
```

Comment:
you are right about on the "belief". It is supposed to check if there is anything to combine for the row key. In many cases there will be only one row for that key and there wouldn't be anything to combine and we wouldn't want to incur the IncrementalIndex overhead for that case.
Note that we are using different GroupingComparator for the combiner.
```
job.setCombinerKeyGroupingComparatorClass(BytesWritable.Comparator.class);
```
which is different from what is used at Reducer, that is
```
 job.setGroupingComparatorClass(SortableBytesGroupingComparator.class);
```

reason for different GroupingComparator is that in the Combiner we only want to get rows for a particular key and merge those into a single InputRow. However, at reducer we want to get all the rows (for all keys) for a particular segment.